### ISSUE TITLE
[Feature]: Masking missing features from scArches loss calculation.

### ISSUE THREAD

**Issue by ehvr20** [scverse/scvi-tools#3792]
**Is your feature request related to a problem? Please describe.**
I have a single cell reference atlas consisting of the integration of several datasets using scVI. I am attempting to project a spatial transcriptomic dataset (1000 gene panel) via scArches/load_query. These investigations show the general poor mapping of spatial transcriptomic cells into the single cell latent space.

**Describe the solution you'd like**
A major candidate for this is that only 1000 genes out of the 4500 reference model input genes exist. Reading further into the implementation, I see that the missing genes expression is simply padded as zeros. As such during the learning of scArches the ELBO loss is heavily swayed by the reconstruction between the predicted profile the artificial zeros, far deviating from the single cell. The feature I would like to request, is the ability to mask missing features from the reconstruction loss calculation and subsequent ELBO computation during scArches training.

Do you foresee this being difficult to implement? Has this already been tried before?

**Comment by ehvr20** [comment]
I implemented a test of this myself by adding a monkey patch to the `vae.loss` method to use a boolean feature mask. Results present zero change to result latent embedding despite changes to loss metric in `SCVI.history`.

I am leaving a note of my implementation below in case anyone is curious and I am closing this issue.

**Monkey patch function**
```python
import torch
import numpy as np

from scvi import REGISTRY_KEYS
from scvi.model import SCVI
from scvi.module.base import LossOutput
from scvi.module._constants import MODULE_KEYS
from scvi.utils import unsupported_if_adata_minified

@unsupported_if_adata_minified
def masked_loss(self, tensors, inference_outputs, generative_outputs, kl_weight=1.0):
    from torch.distributions import kl_divergence
    
    x = tensors[REGISTRY_KEYS.X_KEY]
    kl_divergence_z = kl_divergence(
        inference_outputs[MODULE_KEYS.QZ_KEY], generative_outputs[MODULE_KEYS.PZ_KEY]
    ).sum(dim=-1)
    if not self.use_observed_lib_size:
        kl_divergence_l = kl_divergence(
            inference_outputs[MODULE_KEYS.QL_KEY], generative_outputs[MODULE_KEYS.PL_KEY]
        ).sum(dim=1)
    else:
        kl_divergence_l = torch.zeros_like(kl_divergence_z)


    log_px = generative_outputs[MODULE_KEYS.PX_KEY].log_prob(x)

    mask = getattr(self, "feature_mask", None)
    if mask is not None:
        log_px = log_px * mask
    
    reconst_loss = -log_px.sum(-1)

    kl_local_for_warmup = kl_divergence_z
    kl_local_no_warmup = kl_divergence_l

    weighted_kl_local = kl_weight * kl_local_for_warmup + kl_local_no_warmup

    loss = torch.mean(reconst_loss + weighted_kl_local)

    return LossOutput(
        loss=loss,
        reconstruction_loss=reconst_loss,
        kl_local={
            MODULE_KEYS.KL_L_KEY: kl_divergence_l,
            MODULE_KEYS.KL_Z_KEY: kl_divergence_z,
        },
    )

def mask_features_from_loss(model, adata, mask_key):
    
    mask = np.asarray(adata.var[mask_key].values)
    
    if not (isinstance(mask, np.ndarray) and mask.dtype == np.bool_):
        raise ValueError(
            f"`{mask_key}` must correspond to a boolean array indicating features to be retained in loss in `adata.var`. "
            f"Got dtype={getattr(mask, 'dtype', type(mask))} instead."
        )
    
    if mask.sum() == 0:
        raise ValueError("Feature mask contains no True values.")       
    
    device = model.module.device
    model.module.register_buffer(
        'feature_mask',
        torch.tensor(mask).float().unsqueeze(0).to(device)
    )
    
    model.module.loss = masked_loss.__get__(
        model.module, 
        model.module.__class__
    )
```

**Example usage**
```python
# Load scVI reference model as normal with query data
vae_query_masked = SCVI.load_query_data(
    adata_query,
    MODEL_PATH
)

# Monkey patch loss function and add feature mask
mask_features_from_loss(
    vae_query_masked,
    adata_query, 
    'IsCaptured'
)

# Train model with query added
vae_query_masked.train(
    accelerator='gpu',
    ... # other kwargs
    plan_kwargs=dict(
        weight_decay=0.0 # as per tutorial
    ),
)

# Get new latent embedding generated from masked loss
adata_query.obsm['X_scvi_masked'] = vae_query_masked.get_latent_representation()
```

**Comment by ori-kron-wis** [comment]
Thanks @ehvr20 , can you emphasize if this addition helped in your case and in what sense? Losses and latent space remained the same? what did it improve exactly

**Comment by ehvr20** [comment]
Apologies for not being more clear. A personal theory of mine was that padding missing features with zeros as is the current scArches implementation would bias the reconstruction loss computation and subsequent ELBO loss, reducing overall quality of transfer learning. My initial look into how to test this theory was very daunting as such I submitted the issue hoping for guidance. Taking another look at it again, I realised the loss function is exposed close to the surface in the API and could be adjusted using a relatively crude monkey patch.

The function above is the monkey patch I used, which simply masked out all reconstruction losses from missing features. I then ran the model with and without masking saving each to `adata.obsm['X_scvi_masked']` and `adata.obsm['X_scvi']` respectively.

**Observations**
- `model.history` shows a reduction in loss values, showing that the monkey patch was effective in altering the loss function (sanity check).
- `np.allclose(adata.obsm['X_scvi_masked'], adata.obsm['X_scvi']']` yielded `True`. Whilst the losses indeed changed, the output embedding between masked and unmasked was identical - as such it safe to assume the gradient of descent was completely unaffected. So the modification has zero effect, neither improving or degrading transfer learning quality.

So my personal theory was incorrect and padding missing values is perfectly safe at least from my testing. Someone who knows the mathematics more may have found that obvious from the beginning.

**TLDR:**
Masking missing features yields no benefit. Default implementation padding with zeros is not an issue :)
